README.md
## Mapillary Tools

Mapillary Tools is a library for processing and uploading images to [Mapillary](https://www.mapillary.com/).

<!--ts-->

* [Quickstart](#quickstart)
* [Requirements](#requirements)
* [Installation](#installation)
* [Video Support](#video-support)
* [Usage](#usage)
* [Advanced Usage](#advanced-usage)
* [Command Specifications](#command-specifications)
* [Camera Specific](#camera-specific)
* [Troubleshooting](#troubleshooting)
* [Development](#development)

<!--te-->

## Quickstart

Download the latest `mapillary_tools` binaries for your platform here: https://github.com/mapillary/mapillary_tools/releases/tag/v0.7.4

See [more installation instructions](#installation) below.

Upload imagery:

```bash
./mapillary_tools process_and_upload --import_path "path/to/images" --user_name "mapillary_username"
```

## Requirements

### User Authentication

To upload images to Mapillary, an account is required and can be created [here](https://www.mapillary.com/signup). When
using the tools for the first time, user authentication is required. You will be prompted to enter your account
credentials. Only Mapillary account credentials (email and password) are valid for authentication
in `mapillary_tools`. Other Mapillary login options (Facebook, OpenStreetMap, Google+, or ArcGIS) are not supported
with `mapillary_tools`.

### Metadata

To upload images to Mapillary, image `GPS` and `capture time` are minimally required. More
information [here](https://help.mapillary.com/hc/en-us/articles/115001717829-Geotagging-images).

## Installation

### Installing via Pip

Python (3.6 and above) and git are required:

```bash
python3 -m pip install --upgrade git+https://github.com/mapillary/mapillary_tools
```

If you see "Permission Denied" error, try run the command above with `sudo`, or install it in your local [virtualenv](#development) (recommended).

### Installing on Android Devices

A command line program such as Termux is required. Installation can be done without root privileges.  The following commands will install Python 3, pip3, git, and all required libraries for mapillary_tools on Termux:

```bash
pkg install python git build-essential libgeos openssl libjpeg-turbo
python3 -m pip install --upgrade pip wheel
python3 -m pip install --upgrade git+https://github.com/mapillary/mapillary_tools
```

Termux must access the device's internal storage to process and upload images. To do this, use the following command:

```bash
termux-setup-storage
```

Finally, on devices running Android 11, using a command line program, mapillary_tools will process images very slowly if they are in shared internal storage during processing. It is advisable to first move images to the command line program’s native directory before running mapillary_tools.
For an example using Termux, if imagery is stored in the folder “Internal storage/DCIM/mapillaryimages" the following command will move that folder from shared storage to Termux:

```bash
mv -v storage/dcim/mapillaryimages mapillaryimages
```

## Video Support

To sample images from videos, you will also need to install `ffmpeg`.

### Windows

To install `ffmpeg` on Windows, follow
these [instructions](http://adaptivesamples.com/how-to-install-ffmpeg-on-windows/).

### macOS

To install `ffmpeg` on macOS use [Homebrew](https://brew.sh). Once you have Homebrew installed, you can install `ffmpeg`
by running:

```bash
brew install ffmpeg
```

### Ubuntu

To install `ffmpeg` on Ubuntu:

```bash
sudo apt install ffmpeg
```

## Usage

All commands are executed with `mapillary_tools`.

### Available commands

To see the available commands, use the following in the command line (for Windows, adjust the command according the
instructions for execution):

```bash
mapillary_tools -h
```

Executable `mapillary_tools` takes the following arguments:

`-h, --help`: Show help and exit

`--advanced`: Use the tools under an advanced level, with additional arguments and commands available

-----

`command`: Use one of the available commands:

- `process`: Process the images including for instance, geotagging and sequence arrangement
- `upload`: Upload images to Mapillary
- `process_and_upload`: A bundled command for `process` and `upload`

-----

See the command specific help for required and optional arguments:

- Show help for `process` command:

```bash
mapillary_tools process -h
```

-----

- Show advanced help for `process` command:

```bash
mapillary_tools process -h --advanced
```

### Examples

For Windows, adjust the commands according the instructions for execution.

#### Process Images

The command below processes all images in the directory and its sub-directories. It will update the images with
Mapillary-specific metadata in the image EXIF for the user with user name `mapillary_user`. It requires that each image
in the directory contains `capture time` and `GPS`. By default, only the Image Description EXIF tag is overwritten and
duplicate images are flagged to be excluded from upload using default thresholds for duplicate distance 0.1 m and
duplicate angle 5°.

```bash
mapillary_tools process --import_path "path/to/images" --user_name "mapillary_username"
```

#### Upload Images

The command below uploads all images in a directory and its sub-directories. It requires Mapillary-specific metadata in
the image EXIF. It works for images that are captured with Mapillary iOS or Android apps or processed with the `process`
command.

 ```bash
mapillary_tools upload --import_path "path/to/images"
```

#### Process and Upload Images

The command below runs `process` and `upload` consecutively for a directory.

```bash
mapillary_tools process_and_upload --import_path "path/to/images" --user_name "mapillary_username"
```

## Advanced Usage

Available commands for advanced usage:

- Video Specific Commands:
    - sample_video
    - video_process
    - video_process_and_upload
- Process Unit Commands:
    - extract_user_data
    - extract_import_meta_data
    - extract_geotag_data
    - extract_sequence_data
    - extract_upload_params
    - exif_insert
- Other Commands:
    - process_csv
    - interpolate
    - authenticate
    - post_process

### Geotag and Upload

- Run process and upload consecutively, while process is reading geotag data from a gpx track. It requires
  that `capture time` information is embedded in the image EXIF. By default geotag data is stored only in the mapillary
  image description, in the EXIF Image Description tag. If you would like the rest of the tags to be overwritten as
  well, for example to be able to place images on the map for testing purposes, you should pass an additional
  argument `--overwrite_all_EXIF_tags` to overwrite all EXIF tags, or in case you only want to overwrite a specific tag,
  like for example the GPS tag, pass argument `--overwrite_EXIF_gps_tag`.

 ```bash
mapillary_tools process --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --geotag_source "gpx" \
    --geotag_source_path "path/to/gpx_file"

mapillary_tools upload --import_path "path/to/images"
```

or

 ```bash
mapillary_tools process_and_upload --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --geotag_source "gpx" \
    --geotag_source_path "path/to/gpx_file"
```

### Keep original images intact and Upload

- To prevent data loss or control versions, the original images can be left intact by specifying the
  flag `--keep_original`. This will result in the edited image being saved in a copy of the original image, instead of
  the original image itself. Copies are saved in `{$import_path/$image_path/}.mapillary/process_images}` and are deleted
  at the start of every processing run.

```bash
mapillary_tools process --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --keep_original

mapillary_tools upload --import_path "path/to/images"
```

or

 ```bash
mapillary_tools process_and_upload --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --keep_original
```

### Derive image direction and Upload

- Derive image direction (image heading or camera angle) based on image latitude and longitude. If images are missing
  direction, the direction is derived automatically, if direction is present, it will be derived and overwritten only if
  the flag `--interpolate directions` is specified.

 ```bash
mapillary_tools process --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --interpolate_directions

mapillary_tools upload --import_path "path/to/images"
```

or

 ```bash
mapillary_tools process_and_upload --advanced --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --interpolate_directions
```

### Video Sampling and Upload

- Sample the video(s) located in `path/to/videos` into the directory `path/to/images`, at a sample interval of 0.5
  seconds and tag the sampled images with `capture time`. Note that the video frames will always be sampled into sub
  directory `.mapillary/sampled_video_frames/"video_import_path"`, whether import path is specified or not. In
  case `import_path` is specified the final path for the sampled video frames will
  be `"import path"/.mapillary/sampled_video_frames/"video_import_path"` and in case `import_path` is not specified, the
  final path for the sampled video frames will be `path/to/.mapillary/sampled_video_frames/"video_import_path"`.

 ```bash
mapillary_tools sample_video --import_path "path/to/images" \
    --video_import_path "path/to/videos" \
    --video_sample_interval 0.5 --advanced 
```

- Sample the video(s) located in `path/to/videos`, at a sample interval of 2 seconds (default value) and tag the
  resulting images with `capture time`. And then process and upload the resulting images for
  user `username_at_mapillary`, specifying a gpx track to be the source of geotag data. Additionally pass
  the `--overwrite_all_EXIF_tags` so the extracted frames have all the tags set beside the Image Description tag.

```bash
mapillary_tools sample_video --video_import_path "path/to/videos" --advanced

mapillary_tools process --advanced --import_path "path/to/.mapillary/sampled_video_frames/video_import_path" \
    --user_name "mapillary_username" \
    --geotag_source "gpx" \
    --geotag_source_path "path/to/gpx_file" \
    --overwrite_all_EXIF_tags

mapillary_tools upload --import_path "path/to/.mapillary/sampled_video_frames/video_import_path"
```

or

```bash
mapillary_tools video_process_and_upload --video_import_path "path/to/videos" \
    --user_name "mapillary_username" \
    --advanced --geotag_source "gpx" \
    --geotag_source_path "path/to/gpx_file" \
    --overwrite_all_EXIF_tags
```

### Process csv

- Insert image capture time and gps data from a csv file, based on filename:

```bash
mapillary_tools process_csv --import_path "path/to/images" \
    --csv_path "path/to/csv_file" \
    --filename_column 1 \
    --timestamp_column 4 \
    --latitude_column 2 \
    --longitude_column 3 \
    --advanced
```

- Insert image capture time and meta data from a csv file based on the order of image file names (in case filename
  column is missing):

```bash
mapillary_tools process_csv --import_path "path/to/images" \
    --csv_path "path/to/csv_file" \
    --timestamp_column 1 \
    --meta_columns "6,7" \
    --meta_names "random_name1,random_name2" \
    --meta_types "double,string" \
    --advanced
```

## Command Specifications

### `process`

The `process` command will format the required and optional meta data into a Mapillary image description and insert it
in the image EXIF Image Description tag. Images are required to contain image capture time, latitude, longitude and
camera direction in the image EXIF. Under advanced usage, additional functionalities are available, for example latitude
and longitude can be read from a gpx track file or a GoPro video, camera direction can be derived based on latitude and
longitude, duplicates can be kept instead of excluded from the upload etc. See the command specific help for required
and optional arguments, add `--advanced` to see additional advanced optional arguments.

#### Examples

- process all images for user `mapillary_user`, in the directory `path/to/images` and its sub-directories:

```bash
mapillary_tools process --import_path "path/to/images" \
    --user_name "mapillary_username"
```

- process all images for user `mapillary_user`, in the directory `path/to/images`, skipping the images in its
  sub-directories, rerunning process for all images that were not already uploaded and printing out extra warnings or
  errors.

```bash
mapillary_tools process --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --verbose \
    --rerun \
    --skip_subfolders
```

#### Advanced Examples

- Process all images for user `mapillary_user`, in the directory `path/to/images` and its sub-directories, reading
  geotag data from a gpx track stored in file `path/to/gpx_file`, specifying an offset of 2 seconds between the camera
  and gps device, ie, camera is 2 seconds ahead of the gps device and specifying to keep duplicates to be uploaded
  instead of flagging images as duplicates in case they are apart by equal or less then the default 0.1 m and differ by
  the camera angle by equal or less than the default 5°. Additionally pass the `--overwrite_EXIF_gps_tag` to overwrite
  values with the values obtained from the gpx track.

```bash
mapillary_tools process --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --advanced \
    --geotag_source "gpx" \
    --geotag_source_path "path/to/gpx_file" \
    --offset_time 2 \
    --keep_duplicates \
    --overwrite_EXIF_gps_tag
```

- Process all images for user `mapillary_user`, in the directory `path/to/images` and its sub-directories, specifying
  the import to be private imagery belonging to an organization with organization username `mapillary_organization`. You
  can find the organization username in your dashboard.

```bash
mapillary_tools process --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --advanced \
    --private \
    --organization_username "mapillary_organization"
```

- Process all images for user `mapillary_user`, in the directory `path/to/images` and its sub-directories, specifying an
  angle offset of 90° for the camera direction and splitting images into sequences of images apart by less than 100
  meters according to image `GPS` and less than 120 seconds according to image `capture time`. Additionally pass
  the `--overwrite_EXIF_direction_tag` to overwrite values with the additional specified offset.

```bash
mapillary_tools process --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --advanced \
    --offset_angle 90 \
    --cutoff_distance 100 \
    --cutoff_time 120 \
    --overwrite_EXIF_direction_tag
```

### `upload`

Images that have been successfully processed or were taken with the Mapillary app will contain the required Mapillary
image description embedded in the image EXIF and can be uploaded with the `upload` command.

The `upload` command will collect all the images in the import path, while checking for duplicate flags, processing and
uploading logs. If image is flagged as duplicate, was logged with failed process or logged as successfully uploaded, it
will not be added to the upload list.

By default, 5 threads upload in parallel and the script retries 50 times upon encountering a failure. These can be
customized by specifying additional arguments `--number_threads` and `--max_attempts` under `--advanced` usage or with
environment variables in the command line:

    NUMBER_THREADS=10
    MAX_ATTEMPTS=100

#### Examples

- upload all images in the directory `path/to/images` and its sub directories:

```bash
mapillary_tools upload --import_path "path/to/images"
```

- upload all images in the directory `path/to/images`, while skipping its sub directories and specifying to upload with
  10 threads and 10 maximum attempts:

```bash
mapillary_tools upload --import_path "path/to/images" \
    --skip_subfolders \
    --number_threads 10 \
    --max_attempts 10 \
    --advanced
```

### `process_and_upload`

`process_and_upload` command will run `process` and `upload` commands consecutively with combined required and optional
arguments.

#### Examples

- process and upload all the images in directory `path/to/images` and its sub-directories for user `mapillary_user`.

```bash
mapillary_tools process_and_upload --import_path "path/to/images" \
    --user_name "mapillary_username"
```

#### Advanced Examples

- Process and upload all the images in directory `path/to/images` and its sub-directories for user `mapillary_user`,
  while specifying duplicate distance and angle threshold so that duplicate images are consecutive images that are less
  than 0.5 meter apart according to image `GPS` and have less than 1° camera angle difference according to image
  direction.

```bash
mapillary_tools process_and_upload --import_path "path/to/images" \
    --user_name "mapillary_username" \
    --verbose \
    --rerun \
    --duplicate_distance 0.5 \
    --duplicate_angle 1 \
    --advanced
```

### `sample_video`

`sample_video` command will sample a video into images and insert `capture time` to the image EXIF. Capture time is

CHANGELOG.MD
## 0.0.0 (2018-05-30)


### Breaking changes

* The library has been rewritten for this initial release. The library usage is different from before. Refer to the documentation in the [README](https://github.com/mapillary/mapillary_tools/blob/mapillary_tools_v2/README.md) for instructions on requirement installation and usage. Due to an incompatible logging procedure, previous tools need to be used to finish uploading any sequences that were partially uploaded with the previous tools.

### Features and improvements
* Improved authentication procedure
* Improved user interface with only one executable that can run several basic tools with basic arguments, as well as advanced tools and/or advanced arguments
* Improved logging procedure
* Simpler installation with pip
* Modules can be imported in Python to enable easier development of custom process and/or upload scripts

### Bug fixes
* Inadequate README [#219](https://github.com/mapillary/mapillary_tools/issues/219), [#180](https://github.com/mapillary/mapillary_tools/issues/180), [#159](https://github.com/mapillary/mapillary_tools/issues/159), [#229](https://github.com/mapillary/mapillary_tools/issues/229), [#226](https://github.com/mapillary/mapillary_tools/issues/226), [#157](https://github.com/mapillary/mapillary_tools/issues/157)
* Removal of obsolete scripts [#162](https://github.com/mapillary/mapillary_tools/issues/162)
* Inadequate logging [#63](https://github.com/mapillary/mapillary_tools/issues/63)


## 0.0.1 (2018-06-04)

### Bug fixes
* Store the subsecond estimations done in sequence processing in case of identical timestamps


## 0.0.2 (2018-06-20)

### Bug fixes
* Update the timestamps list in case duplicates are flagged


## 0.1.0 (2018-06-12)

### Breaking changes

* Logging was modified to store log files in a hidden directory where the image is located and not in the import path.

### Features and improvements
* DONE file is created in the import path and not where the tools are being run from.
* exifread version requirement upgraded to resolve installation issues
* support geotagging from a csv file
* support various interpolations
* support re-authentication as a command
* support option to keep the original images intact
* print out version
* increase the amount of information printed out without the verbose flag.


## 0.1.4 (2018-08-14)

### Bug fixes
* Create .mapillary folder for image copies in case of passing `--keep_original`
* Add missing global variable EPOCH in interpolation

### Features and improvements
* Do not require filename column to be passed when processing csv, but align csv data and images in order by image file names if filename column is missing
* Support partial matching between image file names and file names in the csv file

## 0.1.5 (2018-08-23)

### Bug fixes
* Do not delete the import path in case it is specified as an import path for video files to be sampled into and that particular video file was already sampled into that import path, but rather issue a warning that user needs to delete existing frames that were sampled in any previous runs

## 0.1.6 (2018-08-29)

### Features and improvements
* Sample video frames into a made up sub directory `.mapillary/sampled_video_frames/"video filename"`, either where the video is located or in the import path if it is specified.
* Add command line arguments for `user_name`, `user_email` and `user_password` for `authenticate` command, in order to avoid the prompt.

## 0.1.7 (2018-09-18)

### Bug fixes
* Fix bug that resulted in sub seconds added twice in case subseconds were written to capture time tag as well as sub second time tag.

## 0.2.0 (2018-09-23)

### Features and improvements
* Additional information and progress bars under verbose in process,
* Partial optimization in DONE file sending,
* Post process command covering process summary, list of image status and options to move image according to status
* Extend support in csv processing
* Adding number of threads and max attempts in upload as command line arguments

### Bug fixes
* Correct the last interpolated direction

## 0.3.0 (2018-10-31)

### Features and improvements
* Enable specification of a time offset in case of already geotagged images
* Add better progress bars and information output in process
* Support Blackvue videos with embedded gps data
* Add a simple `download` command to download all the blurred imaged from Mapillary for a certain `import_path`
* Support import of multiple videos

### Breaking changes
* Argument `--video_file` was renamed to `--video_import_path` as directories of one or more videos can now be processed and uploaded. Note that even if one video is to be processed and uploaded, the directory of the video should be specified as the video import path and not the video file itself.
* Only the Image Description EXIF tag is overwritten with the mapillary image description which includes all the data obtained during the `process` command. If one would like the rest of the tags to be overwritten, an additional argument needs to be passed under advanced usage. Single specific tags can also be overwritten by additional specific corresponding arguments.

## 0.3.1 (2018-12-10)

### Features and improvements
* mapillary_tools available to be downloaded and installed as a package

## 0.4.0 (2019-01-10)

### Features and improvements
* Improved upload stability
* Improved processing of Blackvue videos with embedded gps data
* Added features for process_csv command
* More options for storing local image file path
* Improved and added features for post_process command

### Breaking changes
* Argument `--video_import_path` can be single video or a directory of videos  
* Duplicate flagging is now done automatically with the default duplicate thresholds. To keep duplicates, argument `--keep_duplicates` must be passed.

### Bug fixes
* Store username in case authenticating manually

## 0.4.1 (2018-01-22)

### Features and improvements

python/process.py
#!/usr/bin/env python

import sys
import os
import uuid
import argparse
import json

import lib.processor as processor

from lib.process_user_properties import process_user_properties
from lib.process_import_meta_properties import process_import_meta_properties
from lib.process_geotag_properties import process_geotag_properties
from lib.process_sequence_properties import process_sequence_properties
from lib.process_upload_params import process_upload_params
from lib.insert_MAPJson import insert_MAPJson

'''
Script for uploading images taken with other cameras than
the Mapillary iOS or Android apps.
It runs in the following steps:
    - Mark images that are potential duplicates and will not be uploaded
    - Group images into sequences based on gps and time
    - Interpolate compass angles for each sequence
    - Add Mapillary tags to the images
The following EXIF tags are required:
-GPSLongitude
-GPSLatitude
-(GPSDateStamp and GPSTimeStamp) or DateTimeOriginal or DateTimeDigitized or DateTime
(assumes Python 2.x, for Python 3.x you need to change some module names)
'''

MAPILLARY_UPLOAD_URL = "https://s3-eu-west-1.amazonaws.com/mapillary.uploads.manual.images"
MAPILLARY_DIRECT_UPLOAD_URL = "https://s3-eu-west-1.amazonaws.com/mapillary.uploads.images"
NUMBER_THREADS = int(os.getenv('NUMBER_THREADS', '2'))


def get_args():
    parser = argparse.ArgumentParser(
        description='Process photos to have them uploaded to Mapillary')
    # path to the import photos
    parser.add_argument('path', help='path to your photos')
    # force rerun process, will rewrite the json and update the processing logs
    parser.add_argument(
        '--rerun', help='rerun the processing', action='store_true')
    # user name for the import
    parser.add_argument("--user_name", help="user name", required=True)
    # sequence level parameters
    parser.add_argument('--cutoff_distance', default=600., type=float,
                        help='maximum gps distance in meters within a sequence')
    parser.add_argument('--cutoff_time', default=60., type=float,
                        help='maximum time interval in seconds within a sequence')
    parser.add_argument('--interpolate_directions',
                        help='perform interploation of directions', action='store_true')
    parser.add_argument('--offset_angle', default=0., type=float,
                        help='offset camera angle (90 for right facing, 180 for rear facing, -90 for left facing)')
    parser.add_argument('--remove_duplicates',
                        help='perform duplicate removal', action='store_true')
    parser.add_argument('--duplicate_distance',
                        help='max distance for two images to be considered duplicates in meters', default=0.1)
    parser.add_argument(
        '--duplicate_angle', help='max angle for two images to be considered duplicates in degrees', default=5)
    # geotagging parameters
    parser.add_argument(
        '--geotag_source', help='Provide the source of date/time and gps information needed for geotagging.', action='store',
        choices=['exif', 'gpx', 'csv', 'json'], default="exif")
    parser.add_argument(
        '--geotag_source_path', help='Provide the path to the file source of date/time and gps information needed for geotagging.', action='store',
        default=None)
    # project level parameters
    parser.add_argument(
        '--project', help="add project name in case validation is required", default=None)
    parser.add_argument(
        '--project_key', help="add project to EXIF (project key)", default=None)
    parser.add_argument('--skip_validate_project',
                        help="do not validate project key or projectd name", action='store_true')
    # import level parameters
    parser.add_argument(
        "--device_make", help="Specify device manufacturer. Note this input has precedence over the input read from the import source file.", default=None)
    parser.add_argument(
        "--device_model", help="Specify device model. Note this input has precedence over the input read from the import source file.", default=None)
    parser.add_argument(
        '--add_file_name', help="Add original file name to EXIF. Note this input has precedence over the input read from the import source file.", action='store_true')
    parser.add_argument(
        '--add_import_date', help="Add import date.", action='store_true')
    parser.add_argument('--orientation', help='Specify the image orientation in degrees. Note this might result in image rotation. Note this input has precedence over the input read from the import source file.',
                        choices=[0, 90, 180, 270], type=int, default=None)
    parser.add_argument(
        "--GPS_accuracy", help="GPS accuracy in meters. Note this input has precedence over the input read from the import source file.", default=None)
    parser.add_argument(
        '--import_meta_source', help='Provide the source of import properties.', action='store',
        choices=['exif', 'json'], default=None)
    parser.add_argument(
        '--import_meta_source_path', help='Provide the path to the file source of import specific information. Note, only JSON format is supported.', action='store',
        default=None)
    # master upload
    parser.add_argument('--master_upload', help='Process images with a master key, note: only used by Mapillary employees',
                        action='store_true', default=False)
    # skip certain steps of processing
    parser.add_argument('--skip_user_processing',
                        help='skip the processing of user properties', action='store_true', default=False)
    parser.add_argument('--skip_import_meta_processing',
                        help='skip the processing of import meta data properties', action='store_true', default=False)
    parser.add_argument('--skip_geotagging', help='skip the geotagging',
                        action='store_true', default=False)
    parser.add_argument('--skip_sequence_processing',
                        help='skip the sequence processing', action='store_true', default=False)
    parser.add_argument('--skip_QC', help='skip the quality check',
                        action='store_true', default=False)
    parser.add_argument('--skip_upload_params_processing',
                        help='skip the upload params processing', action='store_true', default=False)
    parser.add_argument('--skip_insert_MAPJson',
                        help='skip the insertion of MAPJsons into image EXIF tag Image Description', action='store_true', default=False)
    # verbose, print out warnings and info
    parser.add_argument(
        '--verbose', help='print debug info', action='store_true', default=False)

    return parser.parse_args()


if __name__ == '__main__':
    '''
    Use from command line as: python processing.py path --user_name
    '''

    if sys.version_info >= (3, 0):
        raise IOError("Incompatible Python version. This script requires Python 2.x, you are using {0}.".format(
            sys.version_info[:2]))

    args = get_args()
    verbose = args.verbose

    # INITIAL SANITY CHECKS ---------------------------------------
    # set import path to images
    import_path = os.path.abspath(args.path)
    # check if it exist and exit if it doesnt
    if not os.path.isdir(import_path):
        print("Error, import directory " + import_path +
              " doesnt not exist, exiting...")
        sys.exit()
    # get the full image list
    full_image_list = []
    for root, dir, files in os.walk(import_path):
        full_image_list.extend(os.path.join(root, file) for file in files if
                               processor.is_image(file))
    # check if any images in the list and exit if none
    if not len(full_image_list):
        print("Error, no images in the import directory " +
              import_path + " or images dont have the extension .jpg, exiting...")
        sys.exit()
    # set user_name
    user_name = args.user_name
    if not user_name:
        print("Error, must provide a valid user name, exiting...")
        sys.exit()
    # ---------------------------------------

    # PROCESS USER PROPERTIES --------------------------------------
    # parameters
    master_upload = args.master_upload
    # function call
    if not args.skip_user_processing:
        process_user_properties(full_image_list, import_path,
                                user_name, master_upload, verbose)
    # ---------------------------------------

    # PROCESS IMPORT PROPERTIES --------------------------------------
    # parameters
    device_make = args.device_make
    device_model = args.device_model
    GPS_accuracy = args.GPS_accuracy
    add_file_name = args.add_file_name
    add_import_date = args.add_import_date
    orientation = args.orientation
    import_meta_source = args.import_meta_source
    import_meta_source_path = args.import_meta_source_path
    # sanity checks
    if import_meta_source_path == None and import_meta_source != None and import_meta_source != "exif":
        print("Error, if reading import properties from external file, rather than image EXIF or command line arguments, you need to provide full path to the log file.")
        sys.exit()
    elif import_meta_source != None and import_meta_source != "exif" and not os.path.isfile(import_meta_source_path):
        print("Error, " + import_meta_source_path + " file source of import properties does not exist. If reading import properties from external file, rather than image EXIF or command line arguments, you need to provide full path to the log file.")
        sys.exit()
    # function call
    if not args.skip_import_meta_processing:
        process_import_meta_properties(full_image_list, import_path, orientation, device_make,
                                       device_model, GPS_accuracy, add_file_name, add_import_date, import_meta_source, import_meta_source_path, verbose)
    # ---------------------------------------

    # PROCESS GEO/TIME PROPERTIES --------------------------------------
    # parameters
    geotag_source = args.geotag_source
    geotag_source_path = args.geotag_source_path
    offset_angle = args.offset_angle
    # sanity checks
    if geotag_source_path == None and geotag_source != "exif":
        # if geotagging from external log file, path to the external log file
        # needs to be provided, if not, exit
        print("Error, if geotagging from external log, rather than image EXIF, you need to provide full path to the log file.")
        sys.exit()
    elif geotag_source != "exif" and not os.path.isfile(geotag_source_path):
        print("Error, " + geotag_source_path +
              " file source of gps/time properties does not exist. If geotagging from external log, rather than image EXIF, you need to provide full path to the log file.")
        sys.exit()
    # function call
    if not args.skip_geotagging:
        process_geotag_properties(
            full_image_list, import_path, geotag_source, geotag_source_path, offset_angle, verbose)
    # ---------------------------------------

    # PROCESS SEQUENCE PROPERTIES --------------------------------------
    # parameters
    cutoff_distance = args.cutoff_distance
    cutoff_time = args.cutoff_time
    interpolate_directions = args.interpolate_directions
    remove_duplicates = args.remove_duplicates
    duplicate_distance = args.duplicate_distance
    duplicate_angle = args.duplicate_angle

    if not args.skip_sequence_processing:
        # function call
        process_sequence_properties(import_path, cutoff_distance, cutoff_time,
                                    interpolate_directions, remove_duplicates, duplicate_distance, duplicate_angle, verbose)
    # ---------------------------------------

    # QC--------------------------------------
    # parameters
    if not args.skip_QC:
        # function call
        pass
    # ---------------------------------------

    # PROCESS UPLOAD PARAMS PROPERTIES --------------------------------------
    # parameters
    if not args.skip_upload_params_processing:
        # function call
        process_upload_params(full_image_list, import_path,
                              user_name, master_upload, verbose)

    # ---------------------------------------

    # COMBINE META DATA AND INSERT INTO EXIF IMAGE DESCRIPTION ---------------
    # function call
    insert_MAPJson(full_image_list, import_path, master_upload,
                   verbose, args.skip_insert_MAPJson)
    # ---------------------------------------
